File: CA_Server/Server/Game/Object/Block.cs
using Google.Protobuf.Protocol;
using System;
using System.Collections.Generic;
using System.Text;
using System.Timers;

namespace Server.Game
{
    public class Block : GameObject
    {
        public Block()
        {
            ObjectType = GameObjectType.Block;
        }

        public override void OnAttacked(GameObject attacker)
        {
            // Pop 패킷 전송
            S_Pop popPacket = new S_Pop();
            popPacket.ObjectId = Id;

            Room.Broadcast(popPacket);

            // 0.5초 후 소멸
            Room.PushAfter(500, Despawn);
        }

        private void Despawn()
        {
            DropItem();

            Room.Push(Room.LeaveGame, Id);
        }

        private void DropItem()
        {
            // TODO : item 생성 (일단 랜덤으로)
            Random rand = new Random();
            int randInt = rand.Next(0, 100);
            int itemType = 0;
            if (randInt < 40)
                return;
            else if (randInt < 60)
                itemType = 1;
            else if (randInt < 75)
                itemType = 2;
            else if (randInt < 80)
                itemType = 3;
            else
                itemType = 4;

            Item item = ObjectManager.Instance.Add<Item>();
            if (item == null)
                return;

            item.ItemType = itemType;
            item.Info.Name = $"Item_{item.ItemType}";
            item.PosInfo.State = CreatureState.Idle;
            item.PosInfo.PosX = PosInfo.PosX;
            item.PosInfo.PosY = PosInfo.PosY;

            GameRoom room = Room;
            room.Push(room.EnterGame, item);
        }
    }
}

File: CA_CppServer/GameServer/Block.cpp
#include "pch.h"
#include "Block.h"
#include "GameRoom.h"
#include "ClientPacketHandler.h"
#include "ObjectManager.h"
#include "Item.h"
#include <random>

void Block::OnAttacked(GameObjectRef attacker)
{
	GameRoomRef gameRoom = room.lock();
	if (gameRoom == nullptr)
		return;

	// Pop 패킷 전송
	Protocol::S_Pop popPacket;
	popPacket.set_objectid(info.objectid());
	auto sendBuffer = ClientPacketHandler::MakeSendBuffer(popPacket);
	gameRoom->DoAsync(&GameRoom::Broadcast, sendBuffer);

	// 0.5초 후 소멸
	gameRoom->DoTimer(500, [=]() { Despawn(); });
}

void Block::Despawn()
{
	GameRoomRef gameRoom = room.lock();
	if (gameRoom == nullptr)
		return;

	DropItem();

	gameRoom->DoAsync(&GameRoom::LeaveGame, info.objectid());
}

void Block::DropItem()
{
	// item 생성 (일단 랜덤으로)
	::random_device rd;
	::mt19937 gen(rd());
	::uniform_int_distribution<int> dis(0, 99);

	int randInt = dis(gen);
	int32 itemType = 0;
	if (randInt < 40)
		return;
	else if (randInt < 60)
		itemType = ItemType::BALLOON;
	else if (randInt < 75)
		itemType = ItemType::POTION;
	else if (randInt < 80)
		itemType = ItemType::POTION_MAKE_POWER_MAX;
	else
		itemType = ItemType::SKATE;

	ItemRef item = make_shared<Item>();
	item->SetObjectId(GObjectManager.GenerateId(Protocol::GameObjectType::ITEM));
	item->itemType = itemType;
	item->info.set_name("Item_" + ::to_string(itemType));
	{
		Protocol::PositionInfo* posInfo = item->info.mutable_posinfo();
		posInfo->set_state(Protocol::CreatureState::IDLE);
		posInfo->set_posx(info.posinfo().posx());
		posInfo->set_posy(info.posinfo().posy());
	}

	room.lock()->DoAsync(&GameRoom::EnterGame, static_pointer_cast<GameObject>(item));
}
